assets/src/design-system/components/typeahead/typeahead.js

```diff
@@ -0,0 +1,350 @@
+      ) {
+        return;
+      }
+
+      if (!isFlexibleValue && activeOption?.label) {
+        inputValue.set(activeOption.label);
+      }
+
+      isOpen.set(false);
+      isMenuFocused.set(false);
+
+      // move focus to next element if available.
+      inputRef.current?.offsetParent?.nextSibling?.focus();
+    },
+    [activeOption, clearId, inputValue, isFlexibleValue, isOpen, isMenuFocused]
+  );
+
+  const handleMenuItemClick = useCallback(
+    (event, menuItem) => {
+      onMenuItemClick?.(event, menuItem);
+      if (!isKeepMenuOpenOnSelection) {
+        handleDismissMenu();
+      }
+    },
+    [handleDismissMenu, isKeepMenuOpenOnSelection, onMenuItemClick]
+  );
+
+  const handleReturnToInput = useCallback(() => inputRef?.current?.focus(), []);
+
+  /**
+   * Callbacks passed to input
+   * handleInputChange: When input changes by user.
+   * handleClearInputValue: When clear button is triggered.
+   * - Will only clear the input value by default.
+   * - If isFlexibleValue and onMenuItemClick are present will also trigger onMenuItemClick with an empty string to clear out selected value (presumably - since this is controlled by parent).
+   * handleTabClearButton: When keyboard user goes from input to clear button.
+   * - Since technically we've already focused out of the input where the menu wants to close, if tab is used on clear button, dismiss menu but don't focus back to input.
+   */
+  const handleInputChange = useCallback(
+    ({ target }) => {
+      inputValue.set(target.value);
+
+      if (target.value.length > 0 && !isOpen.value) {
+        isOpen.set(!isOpen.value);
+      }
+    },
+    [inputValue, isOpen]
+  );
+
+  const handleClearInputValue = useCallback(() => {
+    inputValue.set('');
+    if (isFlexibleValue) {
+      onMenuItemClick?.(null, '');
+    }
+    handleReturnToInput();
+  }, [handleReturnToInput, inputValue, isFlexibleValue, onMenuItemClick]);
+
+  const handleTabClearButton = useCallback(() => {
+    isOpen.set(false);
+    isMenuFocused.set(false);
+
+    if (!isFlexibleValue && activeOption?.label) {
+      inputValue.set(activeOption.label);
+    }
+  }, [isOpen, isMenuFocused, inputValue, isFlexibleValue, activeOption]);
+
+  const focusSentToList = useCallback(() => isMenuFocused.set(true), [
+    isMenuFocused,
+  ]);
+
+  const setInputValueToActiveOption = useCallback(
+    () => activeOption?.label && inputValue.set(activeOption.label),
+    [activeOption, inputValue]
+  );
+
+  /**
+   * handleInputKeyPress
+   * Watch keyDown of input for keys that change interaction
+   * ESC: Close menu
+   * - Also reset input value to active option if isFlexibleValue is false on escape.
+   * TAB:
+   * - Close menu on tab if inputValue length is 0
+   * - And reset input value to active option if not isFlexibleValue
+   * ARROW DOWN: Send focus to menu list
+   * ENTER: Trigger menuItemClick if isFlexibleValue is true
+   */
```

Comment:
This is really helpful

Reply:
i simplified this and removed the note, let me know if you feel like it should still be there.